// @fleet-sdk/core/build/lib/main/builder/outputBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OutputBuilder = exports.SAFE_MIN_BOX_VALUE = void 0;
const invalidRegistersPacking_1 = require("../errors/invalidRegistersPacking");
const undefinedCreationHeight_1 = require("../errors/undefinedCreationHeight");
const undefinedMintingContext_1 = require("../errors/undefinedMintingContext");
const models_1 = require("../models");
const tokensCollection_1 = require("../models/collections/tokensCollection");
const byteColl_1 = require("../serialization/sigma/byteColl");
const arrayUtils_1 = require("../utils/arrayUtils");
const bigIntUtils_1 = require("../utils/bigIntUtils");
const boxUtils_1 = require("../utils/boxUtils");
const objectUtils_1 = require("../utils/objectUtils");
const stringUtils_1 = require("../utils/stringUtils");
exports.SAFE_MIN_BOX_VALUE = 1000000n;
class OutputBuilder {
    _value;
    _address;
    _tokens;
    _creationHeight;
    _registers;
    _minting;
    constructor(value, recipient, creationHeight) {
        this._value = (0, bigIntUtils_1.toBigInt)(value);
        this._creationHeight = creationHeight;
        this._tokens = new tokensCollection_1.TokensCollection();
        this._registers = {};
        if (typeof recipient === "string") {
            this._address = (0, stringUtils_1.isHex)(recipient)
                ? models_1.ErgoAddress.fromErgoTree(recipient)
                : models_1.ErgoAddress.fromBase58(recipient);
        }
        else {
            this._address = recipient;
        }
    }
    get value() {
        return this._value;
    }
    get address() {
        return this._address;
    }
    get ergoTree() {
        return this._address.ergoTree;
    }
    get creationHeight() {
        return this._creationHeight;
    }
    get tokens() {
        return this._tokens;
    }
    get additionalRegisters() {
        return this._registers;
    }
    get minting() {
        return this._minting;
    }
    addTokens(tokens, options) {
        if (tokens instanceof tokensCollection_1.TokensCollection) {
            this._tokens.add(tokens.toArray(), options);
        }
        else {
            this._tokens.add(tokens, options);
        }
        return this;
    }
    mintToken(token) {
        this._minting = { ...token, amount: (0, bigIntUtils_1.toBigInt)(token.amount) };
        return this;
    }
    setCreationHeight(height, options) {
        if ((0, objectUtils_1.isUndefined)(options) ||
            options.replace === true ||
            (options.replace === false && (0, objectUtils_1.isUndefined)(this._creationHeight))) {
            this._creationHeight = height;
        }
        return this;
    }
    setAdditionalRegisters(registers) {
        this._registers = (0, objectUtils_1.removeUndefined)(registers);
        if (!(0, boxUtils_1.areRegistersDenselyPacked)(registers)) {
            throw new invalidRegistersPacking_1.InvalidRegistersPacking();
        }
        return this;
    }
    eject(ejector) {
        ejector({ tokens: this._tokens });
    }
    build(transactionInputs) {
        let tokens = this.tokens.toArray();
        if (this.minting) {
            if ((0, arrayUtils_1.isEmpty)(transactionInputs)) {
                throw new undefinedMintingContext_1.UndefinedMintingContext();
            }
            if ((0, arrayUtils_1.isEmpty)(this.additionalRegisters)) {
                this.setAdditionalRegisters({
                    R4: new byteColl_1.ByteColl(Buffer.from(this.minting.name || "", "utf-8")).toString(),
                    R5: new byteColl_1.ByteColl(Buffer.from(this.minting.description || "", "utf-8")).toString(),
                    R6: new byteColl_1.ByteColl(Buffer.from(this.minting.decimals?.toString() || "0", "utf-8")).toString()
                });
            }
            tokens = [
                {
                    tokenId: (0, arrayUtils_1.first)(transactionInputs).boxId,
                    amount: this.minting.amount
                },
                ...tokens
            ];
        }
        if ((0, objectUtils_1.isUndefined)(this.creationHeight)) {
            throw new undefinedCreationHeight_1.UndefinedCreationHeight();
        }
        return {
            value: this.value.toString(),
            ergoTree: this.ergoTree,
            creationHeight: this.creationHeight,
            assets: tokens.map((token) => {
                return {
                    tokenId: token.tokenId,
                    amount: token.amount.toString()
                };
            }),
            additionalRegisters: this.additionalRegisters
        };
    }
}
exports.OutputBuilder = OutputBuilder;
