// @fleet-sdk/core/build/lib/module/utils/arrayUtils.js
export function isEmpty(obj) {
    if (!obj) {
        return true;
    }
    return Array.isArray(obj) ? obj.length === 0 : Object.keys(obj).length === 0;
}
export function some(obj) {
    return !isEmpty(obj);
}
export function first(array) {
    if (!array) {
        return;
    }
    if (!array[0]) {
        throw Error("Empty array.");
    }
    return array[0];
}
/**
 * Check for duplicate elements using the equality operator
 */
export function hasDuplicates(array) {
    return array.some((item, index) => {
        return array.indexOf(item) !== index;
    });
}
/**
 * Check for duplicate keys in complex elements
 */
export function hasDuplicatesBy(array, selector) {
    return array.some((item, index) => {
        return array.findIndex((x) => selector(x) === selector(item)) !== index;
    });
}
export function chunk(array, size) {
    if (array.length <= size) {
        return [array];
    }
    const chunks = [];
    for (let i = 0; i < array.length; i += size) {
        chunks.push(array.slice(i, i + size));
    }
    return chunks;
}
export function orderBy(array, iteratee, order = "asc") {
    return array.sort((a, b) => {
        if (iteratee(a) > iteratee(b)) {
            return order === "asc" ? 1 : -1;
        }
        else if (iteratee(a) < iteratee(b)) {
            return order === "asc" ? -1 : 1;
        }
        else {
            return 0;
        }
    });
}
