// @fleet-sdk/core/build/lib/module/models/collections/outputsCollection.js
import { NotFoundError } from "../../errors";
import { some } from "../../utils/arrayUtils";
import { isUndefined } from "../../utils/objectUtils";
import { Collection } from "./collection";
export class OutputsCollection extends Collection {
    constructor(outputs) {
        super();
        if (outputs) {
            this.add(outputs);
        }
    }
    _add(outputBuilder) {
        this._items.push(outputBuilder);
    }
    add(outputs) {
        if (!Array.isArray(outputs)) {
            this._add(outputs);
            return this;
        }
        for (const output of outputs) {
            this._add(output);
        }
        return this;
    }
    remove(outputs) {
        let index = -1;
        if (typeof outputs === "number") {
            if (this._isIndexOutOfBounds(outputs)) {
                throw new RangeError(`Index '${outputs}' is out of range.`);
            }
            index = outputs;
        }
        else {
            index = this._items.lastIndexOf(outputs);
            if (this._isIndexOutOfBounds(index)) {
                throw new NotFoundError("The output you are trying to remove is not present in the outputs collection.");
            }
        }
        if (index > -1) {
            this._items.splice(index, 1);
        }
        return this;
    }
    clone() {
        return new OutputsCollection(this._items);
    }
    sum(basis) {
        const tokens = {};
        let nanoErgs = 0n;
        if (basis) {
            if (basis.nanoErgs) {
                nanoErgs = basis.nanoErgs;
            }
            if (some(basis.tokens)) {
                for (const token of basis.tokens) {
                    if (isUndefined(token.amount)) {
                        continue;
                    }
                    tokens[token.tokenId] = (tokens[token.tokenId] || 0n) + token.amount;
                }
            }
        }
        for (const box of this._items) {
            nanoErgs += box.value;
            for (const token of box.tokens) {
                tokens[token.tokenId] = (tokens[token.tokenId] || 0n) + token.amount;
            }
        }
        return {
            nanoErgs,
            tokens: Object.keys(tokens).map((tokenId) => ({ tokenId, amount: tokens[tokenId] }))
        };
    }
}
