// @fleet-sdk/core/build/lib/module/builder/selector/boxSelector.js
import { DuplicateInputSelectionError } from "../../errors/duplicateInputSelectionError";
import { InsufficientInputs } from "../../errors/insufficientInputs";
import { InputsCollection } from "../../models";
import { hasDuplicatesBy, isEmpty, orderBy, some } from "../../utils/arrayUtils";
import { sumBy, toBigInt } from "../../utils/bigIntUtils";
import { sumByTokenId } from "../../utils/boxUtils";
import { isDefined } from "../../utils/objectUtils";
import { AccumulativeSelectionStrategy } from "./strategies/accumulativeSelectionStrategy";
import { CustomSelectionStrategy } from "./strategies/customSelectionStrategy";
export class BoxSelector {
    _inputs;
    _strategy;
    _ensureFilterPredicate;
    _inputsSortSelector;
    _inputsSortDir;
    constructor(inputs) {
        if (inputs instanceof InputsCollection) {
            this._inputs = inputs.toArray();
        }
        else {
            this._inputs = inputs;
        }
    }
    defineStrategy(strategy) {
        if (this._isISelectionStrategyImplementation(strategy)) {
            this._strategy = strategy;
        }
        else {
            this._strategy = new CustomSelectionStrategy(strategy);
        }
        return this;
    }
    select(target) {
        if (!this._strategy) {
            this._strategy = new AccumulativeSelectionStrategy();
        }
        const remaining = { ...target };
        let unselected = [...this._inputs];
        let selected;
        if (isDefined(this._ensureFilterPredicate)) {
            const predicate = this._ensureFilterPredicate;
            selected = unselected.filter(predicate);
            unselected = unselected.filter((input) => !predicate(input));
            if (isDefined(remaining.nanoErgs)) {
                remaining.nanoErgs -= sumBy(selected, (input) => input.value);
            }
            if (isDefined(remaining.tokens) && selected.some((input) => !isEmpty(input.assets))) {
                remaining.tokens.forEach((tokenTarget) => {
                    if (tokenTarget.amount) {
                        tokenTarget.amount -= sumByTokenId(selected, tokenTarget.tokenId);
                    }
                });
            }
        }
        else {
            selected = [];
        }
        unselected = this._sort(unselected);
        selected = selected.concat(this._strategy.select(unselected, remaining));
        if (hasDuplicatesBy(selected, (item) => item.boxId)) {
            throw new DuplicateInputSelectionError();
        }
        const unreached = this._getUnreachedTargets(selected, target);
        if (some(unreached)) {
            throw new InsufficientInputs(unreached);
        }
        return selected;
    }
    _getUnreachedTargets(inputs, target) {
        const unreached = {};
        const selectedNanoergs = sumBy(inputs, (input) => input.value);
        if (target.nanoErgs && target.nanoErgs > selectedNanoergs) {
            unreached["nanoErgs"] = target.nanoErgs - selectedNanoergs;
        }
        if (isEmpty(target.tokens)) {
            return unreached;
        }
        for (const tokenTarget of target.tokens) {
            const totalSelected = sumByTokenId(inputs, tokenTarget.tokenId);
            if (isDefined(tokenTarget.amount) && tokenTarget.amount > totalSelected) {
                unreached[tokenTarget.tokenId] = tokenTarget.amount - totalSelected;
            }
        }
        return unreached;
    }
    _sort(inputs) {
        if (!this._inputsSortSelector) {
            return orderBy(inputs, (input) => input.creationHeight, "asc");
        }
        return orderBy(inputs, this._inputsSortSelector, this._inputsSortDir || "asc");
    }
    ensureInclusion(predicate) {
        this._ensureFilterPredicate = predicate;
        return this;
    }
    orderBy(selector, direction) {
        this._inputsSortSelector = selector;
        this._inputsSortDir = direction;
        return this;
    }
    _isISelectionStrategyImplementation(obj) {
        if (obj.select) {
            return true;
        }
        return false;
    }
    static buildTargetFrom(boxes) {
        const tokens = {};
        let nanoErgs = 0n;
        for (const box of boxes) {
            nanoErgs += toBigInt(box.value);
            for (const token of box.assets) {
                tokens[token.tokenId] = (tokens[token.tokenId] || 0n) + toBigInt(token.amount);
            }
        }
        return {
            nanoErgs,
            tokens: Object.keys(tokens).map((tokenId) => ({ tokenId, amount: tokens[tokenId] }))
        };
    }
}
