// @fleet-sdk/core/build/lib/main/utils/arrayUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.orderBy = exports.chunk = exports.hasDuplicatesBy = exports.hasDuplicates = exports.first = exports.some = exports.isEmpty = void 0;
function isEmpty(obj) {
    if (!obj) {
        return true;
    }
    return Array.isArray(obj) ? obj.length === 0 : Object.keys(obj).length === 0;
}
exports.isEmpty = isEmpty;
function some(obj) {
    return !isEmpty(obj);
}
exports.some = some;
function first(array) {
    if (!array) {
        return;
    }
    if (!array[0]) {
        throw Error("Empty array.");
    }
    return array[0];
}
exports.first = first;
/**
 * Check for duplicate elements using the equality operator
 */
function hasDuplicates(array) {
    return array.some((item, index) => {
        return array.indexOf(item) !== index;
    });
}
exports.hasDuplicates = hasDuplicates;
/**
 * Check for duplicate keys in complex elements
 */
function hasDuplicatesBy(array, selector) {
    return array.some((item, index) => {
        return array.findIndex((x) => selector(x) === selector(item)) !== index;
    });
}
exports.hasDuplicatesBy = hasDuplicatesBy;
function chunk(array, size) {
    if (array.length <= size) {
        return [array];
    }
    const chunks = [];
    for (let i = 0; i < array.length; i += size) {
        chunks.push(array.slice(i, i + size));
    }
    return chunks;
}
exports.chunk = chunk;
function orderBy(array, iteratee, order = "asc") {
    return array.sort((a, b) => {
        if (iteratee(a) > iteratee(b)) {
            return order === "asc" ? 1 : -1;
        }
        else if (iteratee(a) < iteratee(b)) {
            return order === "asc" ? -1 : 1;
        }
        else {
            return 0;
        }
    });
}
exports.orderBy = orderBy;
