// @fleet-sdk/core/build/lib/main/utils/boxUtils.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mockBoxes_1 = require("../mocks/mockBoxes");
const bigIntUtils_1 = require("./bigIntUtils");
const boxUtils_1 = require("./boxUtils");
describe("Box sumByTokenId", () => {
    it("Should sum correctly", () => {
        const inputs = mockBoxes_1.regularBoxesMock.filter((input) => [
            "2555e34138d276905fe0bc19240bbeca10f388a71f7b4d2f65a7d0bfd23c846d",
            "467b6867c6726cc5484be3cbddbf55c30c0a71594a20c1ac28d35b5049632444",
            "2555e34138d276905fe0bc19240bbeca10f388a71f7b4d2f65a7d0bfd23c846d"
        ].includes(input.boxId));
        expect((0, boxUtils_1.sumByTokenId)(inputs, "0cd8c9f416e5b1ca9f986a7f10a84191dfb85941619e49e53c0dc30ebf83324b")).toBe(3819n);
    });
    it("Should return zero for empty arrays", () => {
        expect((0, boxUtils_1.sumByTokenId)([], "0cd8c9f416e5b1ca9f986a7f10a84191dfb85941619e49e53c0dc30ebf83324b")).toBe(0n);
    });
});
describe("Sum boxes", () => {
    it("Should sum all tokens and nanoErgs", () => {
        const boxes = mockBoxes_1.regularBoxesMock.filter((input) => [
            "e56847ed19b3dc6b72828fcfb992fdf7310828cf291221269b7ffc72fd66706e",
            "a2c9821f5c2df9c320f17136f043b33f7716713ab74c84d687885f9dd39d2c8a",
            "3e67b4be7012956aa369538b46d751a4ad0136138760553d5400a10153046e52",
            "2555e34138d276905fe0bc19240bbeca10f388a71f7b4d2f65a7d0bfd23c846d"
        ].includes(input.boxId));
        expect((0, boxUtils_1.sumBoxes)(boxes)).toEqual({
            nanoErgs: (0, bigIntUtils_1.sumBy)(boxes, (x) => x.value),
            tokens: [
                {
                    tokenId: "007fd64d1ee54d78dd269c8930a38286caa28d3f29d27cadcb796418ab15c283",
                    amount: 226652336n
                },
                {
                    tokenId: "0cd8c9f416e5b1ca9f986a7f10a84191dfb85941619e49e53c0dc30ebf83324b",
                    amount: 10n
                }
            ]
        });
    });
    it("Should sum if box doesn't contains tokens", () => {
        const boxes = mockBoxes_1.regularBoxesMock.filter((input) => [
            "e56847ed19b3dc6b72828fcfb992fdf7310828cf291221269b7ffc72fd66706e",
            "30cb07d93f16f5b052e9f56c1b5dfb83db9ccaeb467dde064933afc23beb6f5f"
        ].includes(input.boxId));
        expect((0, boxUtils_1.sumBoxes)(boxes)).toEqual({
            nanoErgs: (0, bigIntUtils_1.sumBy)(boxes, (x) => x.value),
            tokens: []
        });
    });
});
describe("Densely pack check - areRegistersDenselyPacked()", () => {
    it("Should pass for VALID registers packing", () => {
        expect((0, boxUtils_1.areRegistersDenselyPacked)({
            R4: "0580c0fc82aa02"
        })).toBeTruthy();
        expect((0, boxUtils_1.areRegistersDenselyPacked)({
            R4: "0580c0fc82aa02",
            R5: "07036b84756b351ee1c57fd8c302e66a1bb927e5d8b6e1a8e085935de3971f84ae17"
        })).toBeTruthy();
        expect((0, boxUtils_1.areRegistersDenselyPacked)({
            R4: "0580c0fc82aa02",
            R5: "0580c0fc82aa02",
            R6: "0580c0fc82aa02",
            R7: "0580c0fc82aa02",
            R8: "0580c0fc82aa02",
            R9: "0580c0fc82aa02"
        })).toBeTruthy();
    });
    it("Should fail for INVALID registers packing", () => {
        // R4 not included
        expect((0, boxUtils_1.areRegistersDenselyPacked)({
            R5: "0580c0fc82aa02"
        })).toBeFalsy();
        // R5 not included
        expect((0, boxUtils_1.areRegistersDenselyPacked)({
            R4: "0580c0fc82aa02",
            R6: "07036b84756b351ee1c57fd8c302e66a1bb927e5d8b6e1a8e085935de3971f84ae17"
        })).toBeFalsy();
        // R4, R5 and R6 not included
        expect((0, boxUtils_1.areRegistersDenselyPacked)({
            R7: "07036b84756b351ee1c57fd8c302e66a1bb927e5d8b6e1a8e085935de3971f84ae17"
        })).toBeFalsy();
        // R5, R6, R7 and R8 not included
        expect((0, boxUtils_1.areRegistersDenselyPacked)({
            R4: "0580c0fc82aa02",
            R9: "07036b84756b351ee1c57fd8c302e66a1bb927e5d8b6e1a8e085935de3971f84ae17"
        })).toBeFalsy();
    });
});
