// @fleet-sdk/core/build/lib/module/builder/outputBuilder.js
import { InvalidRegistersPacking } from "../errors/invalidRegistersPacking";
import { UndefinedCreationHeight } from "../errors/undefinedCreationHeight";
import { UndefinedMintingContext } from "../errors/undefinedMintingContext";
import { ErgoAddress } from "../models";
import { TokensCollection } from "../models/collections/tokensCollection";
import { ByteColl } from "../serialization/sigma/byteColl";
import { first, isEmpty } from "../utils/arrayUtils";
import { toBigInt } from "../utils/bigIntUtils";
import { areRegistersDenselyPacked } from "../utils/boxUtils";
import { isUndefined, removeUndefined } from "../utils/objectUtils";
import { isHex } from "../utils/stringUtils";
export const SAFE_MIN_BOX_VALUE = 1000000n;
export class OutputBuilder {
    _value;
    _address;
    _tokens;
    _creationHeight;
    _registers;
    _minting;
    constructor(value, recipient, creationHeight) {
        this._value = toBigInt(value);
        this._creationHeight = creationHeight;
        this._tokens = new TokensCollection();
        this._registers = {};
        if (typeof recipient === "string") {
            this._address = isHex(recipient)
                ? ErgoAddress.fromErgoTree(recipient)
                : ErgoAddress.fromBase58(recipient);
        }
        else {
            this._address = recipient;
        }
    }
    get value() {
        return this._value;
    }
    get address() {
        return this._address;
    }
    get ergoTree() {
        return this._address.ergoTree;
    }
    get creationHeight() {
        return this._creationHeight;
    }
    get tokens() {
        return this._tokens;
    }
    get additionalRegisters() {
        return this._registers;
    }
    get minting() {
        return this._minting;
    }
    addTokens(tokens, options) {
        if (tokens instanceof TokensCollection) {
            this._tokens.add(tokens.toArray(), options);
        }
        else {
            this._tokens.add(tokens, options);
        }
        return this;
    }
    mintToken(token) {
        this._minting = { ...token, amount: toBigInt(token.amount) };
        return this;
    }
    setCreationHeight(height, options) {
        if (isUndefined(options) ||
            options.replace === true ||
            (options.replace === false && isUndefined(this._creationHeight))) {
            this._creationHeight = height;
        }
        return this;
    }
    setAdditionalRegisters(registers) {
        this._registers = removeUndefined(registers);
        if (!areRegistersDenselyPacked(registers)) {
            throw new InvalidRegistersPacking();
        }
        return this;
    }
    eject(ejector) {
        ejector({ tokens: this._tokens });
    }
    build(transactionInputs) {
        let tokens = this.tokens.toArray();
        if (this.minting) {
            if (isEmpty(transactionInputs)) {
                throw new UndefinedMintingContext();
            }
            if (isEmpty(this.additionalRegisters)) {
                this.setAdditionalRegisters({
                    R4: new ByteColl(Buffer.from(this.minting.name || "", "utf-8")).toString(),
                    R5: new ByteColl(Buffer.from(this.minting.description || "", "utf-8")).toString(),
                    R6: new ByteColl(Buffer.from(this.minting.decimals?.toString() || "0", "utf-8")).toString()
                });
            }
            tokens = [
                {
                    tokenId: first(transactionInputs).boxId,
                    amount: this.minting.amount
                },
                ...tokens
            ];
        }
        if (isUndefined(this.creationHeight)) {
            throw new UndefinedCreationHeight();
        }
        return {
            value: this.value.toString(),
            ergoTree: this.ergoTree,
            creationHeight: this.creationHeight,
            assets: tokens.map((token) => {
                return {
                    tokenId: token.tokenId,
                    amount: token.amount.toString()
                };
            }),
            additionalRegisters: this.additionalRegisters
        };
    }
}
