// @fleet-sdk/core/build/lib/main/builder/selector/boxSelector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BoxSelector = void 0;
const duplicateInputSelectionError_1 = require("../../errors/duplicateInputSelectionError");
const insufficientInputs_1 = require("../../errors/insufficientInputs");
const models_1 = require("../../models");
const arrayUtils_1 = require("../../utils/arrayUtils");
const bigIntUtils_1 = require("../../utils/bigIntUtils");
const boxUtils_1 = require("../../utils/boxUtils");
const objectUtils_1 = require("../../utils/objectUtils");
const accumulativeSelectionStrategy_1 = require("./strategies/accumulativeSelectionStrategy");
const customSelectionStrategy_1 = require("./strategies/customSelectionStrategy");
class BoxSelector {
    _inputs;
    _strategy;
    _ensureFilterPredicate;
    _inputsSortSelector;
    _inputsSortDir;
    constructor(inputs) {
        if (inputs instanceof models_1.InputsCollection) {
            this._inputs = inputs.toArray();
        }
        else {
            this._inputs = inputs;
        }
    }
    defineStrategy(strategy) {
        if (this._isISelectionStrategyImplementation(strategy)) {
            this._strategy = strategy;
        }
        else {
            this._strategy = new customSelectionStrategy_1.CustomSelectionStrategy(strategy);
        }
        return this;
    }
    select(target) {
        if (!this._strategy) {
            this._strategy = new accumulativeSelectionStrategy_1.AccumulativeSelectionStrategy();
        }
        const remaining = { ...target };
        let unselected = [...this._inputs];
        let selected;
        if ((0, objectUtils_1.isDefined)(this._ensureFilterPredicate)) {
            const predicate = this._ensureFilterPredicate;
            selected = unselected.filter(predicate);
            unselected = unselected.filter((input) => !predicate(input));
            if ((0, objectUtils_1.isDefined)(remaining.nanoErgs)) {
                remaining.nanoErgs -= (0, bigIntUtils_1.sumBy)(selected, (input) => input.value);
            }
            if ((0, objectUtils_1.isDefined)(remaining.tokens) && selected.some((input) => !(0, arrayUtils_1.isEmpty)(input.assets))) {
                remaining.tokens.forEach((tokenTarget) => {
                    if (tokenTarget.amount) {
                        tokenTarget.amount -= (0, boxUtils_1.sumByTokenId)(selected, tokenTarget.tokenId);
                    }
                });
            }
        }
        else {
            selected = [];
        }
        unselected = this._sort(unselected);
        selected = selected.concat(this._strategy.select(unselected, remaining));
        if ((0, arrayUtils_1.hasDuplicatesBy)(selected, (item) => item.boxId)) {
            throw new duplicateInputSelectionError_1.DuplicateInputSelectionError();
        }
        const unreached = this._getUnreachedTargets(selected, target);
        if ((0, arrayUtils_1.some)(unreached)) {
            throw new insufficientInputs_1.InsufficientInputs(unreached);
        }
        return selected;
    }
    _getUnreachedTargets(inputs, target) {
        const unreached = {};
        const selectedNanoergs = (0, bigIntUtils_1.sumBy)(inputs, (input) => input.value);
        if (target.nanoErgs && target.nanoErgs > selectedNanoergs) {
            unreached["nanoErgs"] = target.nanoErgs - selectedNanoergs;
        }
        if ((0, arrayUtils_1.isEmpty)(target.tokens)) {
            return unreached;
        }
        for (const tokenTarget of target.tokens) {
            const totalSelected = (0, boxUtils_1.sumByTokenId)(inputs, tokenTarget.tokenId);
            if ((0, objectUtils_1.isDefined)(tokenTarget.amount) && tokenTarget.amount > totalSelected) {
                unreached[tokenTarget.tokenId] = tokenTarget.amount - totalSelected;
            }
        }
        return unreached;
    }
    _sort(inputs) {
        if (!this._inputsSortSelector) {
            return (0, arrayUtils_1.orderBy)(inputs, (input) => input.creationHeight, "asc");
        }
        return (0, arrayUtils_1.orderBy)(inputs, this._inputsSortSelector, this._inputsSortDir || "asc");
    }
    ensureInclusion(predicate) {
        this._ensureFilterPredicate = predicate;
        return this;
    }
    orderBy(selector, direction) {
        this._inputsSortSelector = selector;
        this._inputsSortDir = direction;
        return this;
    }
    _isISelectionStrategyImplementation(obj) {
        if (obj.select) {
            return true;
        }
        return false;
    }
    static buildTargetFrom(boxes) {
        const tokens = {};
        let nanoErgs = 0n;
        for (const box of boxes) {
            nanoErgs += (0, bigIntUtils_1.toBigInt)(box.value);
            for (const token of box.assets) {
                tokens[token.tokenId] = (tokens[token.tokenId] || 0n) + (0, bigIntUtils_1.toBigInt)(token.amount);
            }
        }
        return {
            nanoErgs,
            tokens: Object.keys(tokens).map((tokenId) => ({ tokenId, amount: tokens[tokenId] }))
        };
    }
}
exports.BoxSelector = BoxSelector;
