// @fleet-sdk/core/build/lib/main/serialization/sigma/sigmaConstructorTypeCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SigmaConstructorTypeCode = void 0;
var SigmaConstructorTypeCode;
(function (SigmaConstructorTypeCode) {
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["EmbeddableType"] = 1] = "EmbeddableType";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Coll"] = 12] = "Coll";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["NestedColl"] = 24] = "NestedColl";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Option"] = 36] = "Option";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["OptionColl"] = 48] = "OptionColl";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Tuple2"] = 60] = "Tuple2";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Tuple3"] = 72] = "Tuple3";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Tuple4"] = 84] = "Tuple4";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["TupleN"] = 96] = "TupleN";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Any"] = 97] = "Any";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Unit"] = 98] = "Unit";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Box"] = 99] = "Box";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["AvlTree"] = 100] = "AvlTree";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Context"] = 101] = "Context";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Header"] = 104] = "Header";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["PreHeader"] = 105] = "PreHeader";
    SigmaConstructorTypeCode[SigmaConstructorTypeCode["Global"] = 106] = "Global";
})(SigmaConstructorTypeCode = exports.SigmaConstructorTypeCode || (exports.SigmaConstructorTypeCode = {}));
